### PR TITLE
[#12] [iOS] [Integrate] [Part3] As a user, I can reset my password

### LINKED ISSUE #12: [iOS] [Integrate] As a user, I can reset my password
## Why

When user forgot their password, they should be able to reset it.

## Acceptance Criteria?

- Show forgot password page when user click `reset` from login page.
- Call endpoint when user press `Reset` on the reset password screen:
`/api/v1/passwords
`
```
{
    "user": {
        "email": "junan.cs@gmail.com"
    },
    "client_id": "{{client_id}}",
    "client_secret": "{{client_secret}}"
}
```

- Show loading indicator.
- When success, show login screen and schedule a push notification.

## Resources

[Postman - Nimble Survey API](https://documenter.getpostman.com/view/11835486/T1LJjTqj#d15a6ecc-89d4-4a4b-956d-618b98a13062)

### PR DESCRIPTION
**Pull request by blyscuit**
- close #12 

## What happened

Add tests to Reset Password feature.

## Insight

- Create `AnyFlow` on KMM to allow mocking flow on iOS.
- Create `Combine` extensions to allow waiting on unit test.
- Add `ResetPasswordUseCase` to Automockable.
 
## Proof Of Work

Test passes. Coverage increases.

Screen Shot 2022-11-30 at 17 24 06

### CODE REVIEW

File: iosApp/tools/sourcery/templates/AutoMockable.stencil [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import Foundation
+import Shared
```
Comment (suho) [inline iosApp/tools/sourcery/templates/AutoMockable.stencil]: For this import, it's better to use `autoMockableImports` in the file `.sourcery.yml`

File: iosApp/Survey/Sources/Supports/Helpers/KMM/AutoMockable/Shared+AutoMockable.swift [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+// sourcery: AutoMockable
+protocol ResetPasswordUseCaseKMM: ResetPasswordUseCase {
+
+    func invoke(email: String) -> Kotlinx_coroutines_coreFlow
+    func invokeNative(email: String)
+        -> (
+            @escaping (String, KotlinUnit) -> KotlinUnit,
+            @escaping (Error?, KotlinUnit) -> KotlinUnit
+        ) -> () -> KotlinUnit
```
Comment (minhnimble) [inline iosApp/Survey/Sources/Supports/Helpers/KMM/AutoMockable/Shared+AutoMockable.swift:20]: Ohhh, this is the first time I see this kind of syntax for the return type of a function 🤔😁

File: iosApp/SurveyUITests/Sources/Screens/ScreenProtocol.swift [hunk 3]
```diff
@@ -72,3 +93,23 @@ extension ScreenProtocol {
         button.tap()
     }
 }
+
+// MARK: Wait
+
+extension ScreenProtocol {
+
+    func wait(
+        timeout: TimeInterval = 10,
```
Comment (minhnimble) [inline iosApp/SurveyUITests/Sources/Screens/ScreenProtocol.swift]: Can we use `.default` here? I noticed we created the `default` constant for `TimeInterval` as `10` already, right? 🤔

Comment (blyscuit) [reply iosApp/SurveyUITests/Sources/Screens/ScreenProtocol.swift]: Updated 👍

File: shared/src/commonMain/kotlin/co/nimblehq/blisskmmic/helpers/extensions/ios/AnyFlow.kt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package co.nimblehq.blisskmmic.helpers.extensions.ios
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+
+// For iOS Unit Test
+class AnyFlow<T>(source: Flow<T>): Flow<T> by source {
+
+    constructor(result: T) : this(flow { emit(result) })
```
Comment (suho) [inline shared/src/commonMain/kotlin/co/nimblehq/blisskmmic/helpers/extensions/ios/AnyFlow.kt:9]: Cool! I did not think about this one. I was trying to mock the data with Swift, and it was not working :D

**Review by suho — APPROVED** [review]

**Review by minhnimble — APPROVED** [review]

**Review by doannimble — APPROVED** [review]
Great work ✋

**Pull request merged**